apps/system/js/audio_channel_manager.js

```diff
@@ -0,0 +1,350 @@
+/* global BaseModule */
+'use strict';
+
+(function() {
+  var NONE = -1;
+  var PLAY_CURRENT_AND_NEW_CHANNEL = 0;
+  var PAUSE_CURRENT_CHANNEL = 1;
+  var PAUSE_THEN_RESUME_CURRENT_CHANNEL = 2;
+  var FADE_OUT_CURRENT_CHANNEL = 3;
+  var FADE_OUT_NEW_CHANNEL = 4;
+  var PAUSE_THEN_VIBRATE_CURRENT_CHANNEL = 5;
+  var VIBRATE_NEW_CHANNEL = 6;
+  var COMPETING_POLICIES = {
+    'normal': {
+      'normal': PAUSE_CURRENT_CHANNEL,
+      'content': PAUSE_CURRENT_CHANNEL,
+      'alarm': PAUSE_CURRENT_CHANNEL,
+      'system': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'ringer': PAUSE_CURRENT_CHANNEL,
+      'telphony': PAUSE_CURRENT_CHANNEL,
+      'notification': FADE_OUT_CURRENT_CHANNEL,
+      'publicNotification': FADE_OUT_CURRENT_CHANNEL
+    },
+    'content': {
+      'normal': PAUSE_THEN_RESUME_CURRENT_CHANNEL,
+      'content': PAUSE_THEN_RESUME_CURRENT_CHANNEL,
+      'alarm': PAUSE_THEN_RESUME_CURRENT_CHANNEL,
+      'system': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'ringer': PAUSE_THEN_RESUME_CURRENT_CHANNEL,
+      'telphony': PAUSE_THEN_RESUME_CURRENT_CHANNEL,
+      'notification': FADE_OUT_CURRENT_CHANNEL,
+      'publicNotification': FADE_OUT_CURRENT_CHANNEL
+    },
+    'alarm': {
+      'normal': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'content': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'alarm': PAUSE_CURRENT_CHANNEL,
+      'system': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'ringer': FADE_OUT_CURRENT_CHANNEL,
+      'telphony': FADE_OUT_CURRENT_CHANNEL,
+      'notification': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'publicNotification': PLAY_CURRENT_AND_NEW_CHANNEL
+    },
+    'system': {
+      'normal': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'content': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'alarm': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'system': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'ringer': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'telphony': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'notification': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'publicNotification': PLAY_CURRENT_AND_NEW_CHANNEL
+    },
+    'ringer': {
+      'normal': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'content': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'alarm': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'system': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'ringer': PAUSE_THEN_VIBRATE_CURRENT_CHANNEL,
+      'telphony': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'notification': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'publicNotification': PLAY_CURRENT_AND_NEW_CHANNEL
+    },
+    'telphony': {
+      'normal': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'content': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'alarm': PAUSE_THEN_VIBRATE_CURRENT_CHANNEL,
+      'system': PLAY_CURRENT_AND_NEW_CHANNEL,
+      'ringer': VIBRATE_NEW_CHANNEL,
+      'telphony': NONE,
```

Comment:
Yes, I think so.

But we are discussing that do we need a new audio channel type called `voip` for that. See it in[1].

[1]: https://bugzilla.mozilla.org/show_bug.cgi?id=1126224